### PR TITLE
[C+B] Implemented ban reason support into /ban. Adds BUKKIT-4285.

### PR DESCRIPTION
**Pull request by ghost**
The Issue:
Bukkit does not have /ban reason support, therefore I have implemented it in this pull request.

Justification:
This pull request implements ban reason support into /ban, so you can type /ban Example ExampleReason, like in vanilla.

Breakdown:
By making a string named banReason, the command will check if the args >= 2 and if not, does nothing at all, if so, it will define the variable as a String that the sender sets. Then, once the ban takes place, it will check if there is a reason. If not, it will say the reason is "Banned by an operator.", if so, it provides the reason set!

Ticket:
https://bukkit.atlassian.net/browse/BUKKIT-4285 (not mine)

Relevent PR's:
This is a completely different way of implementing ban reason support, so incase https://github.com/Bukkit/Bukkit/pull/935 won't get merged I've made this one.

https://github.com/Bukkit/CraftBukkit/pull/1323 for the CraftBukkit PR to make the new ban API work.

I apologize if I'm not doing this correctly, this is my first pull request.

### CODE REVIEW

File: src/main/java/org/bukkit/command/defaults/BanCommand.java [hunk 1]
```diff
@@ -27,12 +29,16 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
...
+        if (args.length >= 2) {
+            ban_reason = StringUtils.join(ArrayUtils.subarray(args, 1, args.length), " ");
+        }
+        
         Bukkit.getOfflinePlayer(args[0]).setBanned(true);
 
         Player player = Bukkit.getPlayer(args[0]);
         if (player != null) {
-            player.kickPlayer("Banned by admin.");
+            player.kickPlayer("Banned by admin." + (ban_reason != null ? ("\nReason: " + ban_reason) : ""));
```
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: Is this the same in vanilla?

Comment (ghost) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: I just checked the mc-dev code from 1.6.4, yes.

Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: What about in 1.7? I doubt it's changed, but just want to make sure

Comment (ghost) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: mc-dev isn't updated for 1.7, but yeah I'm sure it wasn't updated.

Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: What is the ban message then... You don't need to check code for this.

I can't test it right now :(

Comment (ghost) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: fine ill make a server and check hold on a sec

Comment (ghost) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: Okay, message with NO reason:
You are banned from this server.

Message when reconnect still banned:
You are banned from this server!
Reason: Banned by an operator.
^default reason

Message with reason once kicked, same thing

Message with reason on reconnect:
You are banned from this server!
Reason: A reason.
^custom reason

Every color is white.

Comment (ghost) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: By the way, one of you might want to change the ban messages, as it shouldn't be "Banned by Admin."

**Comment by feildmaster** [comment]
This doesn't actually set the reason the same way as vanilla does it... but it's an interesting work around.

**Comment by ghost** [comment]
Well I guess it can be one of the differences in the Bukkit vanilla commands.

**Comment by ryantheleach** [comment]
Before someone else tells you to, read https://github.com/Bukkit/Bukkit/blob/master/CONTRIBUTING.md and make the changes necessary. It will just save time for everyone.

**Comment by ghost** [comment]
Not sure what's wrong...

**Comment by ryantheleach** [comment]
specifically? https://github.com/Bukkit/Bukkit/blob/master/CONTRIBUTING.md#pull-request-formatting-expectations

However I haven't looked at the code yet as the pull request formatting was the first thing I saw.

**Comment by ghost** [comment]
Okay, added lot's more stuff into the PR. Also I'm gonna fix the string error.

File: src/main/java/org/bukkit/command/defaults/BanCommand.java [hunk 2]
```diff
@@ -27,12 +29,16 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
             return false;
         }
 
-        // TODO: Ban Reason support
+        String ban_reason = null;
```
Comment (Ribesg) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: Should be **banReason** and not **ban_reason** I think.

Comment (ghost) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: Gonna wait for a Bukkit staff member to confirm that.

Comment (Black-Hole) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: https://github.com/Bukkit/Bukkit/blob/master/CONTRIBUTING.md states to follow http://www.oracle.com/technetwork/java/javase/documentation/codeconvtoc-136057.html

Vanilla ban command also saves the ban reason, so maybe someone could open other pull requests for setting and getting the reason in OfflinePlayer.

Comment (gravitylow) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: You should definitely be using camelCase here.

Comment (ghost) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: Yeah... lot's find the current ban API and they either make their own API using the NMS ban classes or use the current API but doing stuff to it, this pull request does the latter.

I'll change the variable now.

Comment (ghost) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: Okay, done.

**Comment by turt2live** [comment]
It would be useful to actually get the ban reason from an API perspective, much like in the PR you've referenced. 

Are you able to do that?

**Comment by ghost** [comment]
Too lazy to add API, because then I would have to make a CraftBukkit pull request too.
Nothing wrong with this, though. It still works.

**Comment by turt2live** [comment]
But accessing the ban reason for plugins is a much needed feature.

**Comment by ghost** [comment]
Fine I'll do it but you owe me a box of cookies :((

**Comment by turt2live** [comment]
I tried and failed 6 times ._.

**Comment by ghost** [comment]
https://github.com/Bukkit/CraftBukkit/pull/1323

File: src/main/java/org/bukkit/OfflinePlayer.java [hunk 3]
```diff
@@ -34,6 +34,14 @@
      * @param banned true if banned
      */
     public void setBanned(boolean banned);
+    
+    /**
+     * Bans or unban this player with reason
+     * 
+     * @param banned true if banned
+     * @param reason for banning player
```
Comment (turt2live) [inline src/main/java/org/bukkit/OfflinePlayer.java]: Should add "ignored if not banning"

Comment (ghost) [inline src/main/java/org/bukkit/OfflinePlayer.java]: ill add it in a sec
By the way, how come you aren't staff on the Bukkit website if you're staff on the Bukkit GitHub?

Comment (turt2live) [inline src/main/java/org/bukkit/OfflinePlayer.java]: ??
img

Comment (ghost) [inline src/main/java/org/bukkit/OfflinePlayer.java]: ah... that must be new
EDIT: I don't see that on the forums... o_o

File: src/main/java/org/bukkit/command/defaults/BanCommand.java [hunk 4]
```diff
@@ -22,13 +24,19 @@ public BanCommand() {
...
             sender.sendMessage(ChatColor.RED + "Usage: " + usageMessage);
             return false;
         }
 
-        // TODO: Ban Reason support
-        Bukkit.getOfflinePlayer(args[0]).setBanned(true);
+        String banReason;
+        if (args.length >= 1) {
+            banReason = StringUtils.join(ArrayUtils.subarray(args, 1, args.length), " ");
+
```
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java:35]: I don't think this blank line is required

File: src/main/java/org/bukkit/OfflinePlayer.java [hunk 5]
```diff
@@ -34,6 +34,14 @@
...
      */
     public void setBanned(boolean banned);
+    
+    /**
+     * Bans or unban this player with reason
+     * 
+     * @param banned true if banned
+     * @param reason for banning player, ignored if not banning
+     */
+    public void setBanned(String reason, boolean banned)
```
Comment (turt2live) [inline src/main/java/org/bukkit/OfflinePlayer.java:44]: Missing a semicolon - does not compile

Comment (turt2live) [inline src/main/java/org/bukkit/OfflinePlayer.java:44]: I also believe that the arguments are in a weird order. Possibly flip them around?

Makes more sense to read `setBanned(true, "no");` than `setBanned("no", true);`

**Comment by turt2live** [comment]
Testing materials: https://drive.google.com/folderview?id=0B6JIYSBkLkAnaTFqUFhZVVR6d0E&usp=sharing&tid=0B6JIYSBkLkAnTTZYd3Z4ZTMzV3c

Drop an item to be banned (note: This plugin does not kick you, so you must disconnect and retry joining the server)

**Comment by turt2live** [comment]
We've decided to pull this a different way, but we appreciate the work put into this!

Thanks!

https://github.com/Bukkit/Bukkit/commit/75427e084d0f2a444375324b0ef7334aa66d93be

**Pull request closed without merging**